File: server/apps/nexotype/schemas/omics/exon_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class ExonCreate(BaseModel):
    """Schema for creating a new exon"""
    transcript_id: int
    ensembl_exon_id: str
    start_position: int
    end_position: int


class ExonUpdate(BaseModel):
    """Schema for updating an exon"""
    transcript_id: int = None
    ensembl_exon_id: str = None
    start_position: int = None
    end_position: int = None


# ==========================================
# Response Schemas
# ==========================================

class ExonDetail(BaseModel):
    """Schema for exon details"""
    id: int
    transcript_id: int
    ensembl_exon_id: str
    start_position: int
    end_position: int
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ExonListResponse(BaseModel):
    """Response schema for listing exons"""
    success: bool
    data: list[ExonDetail] | None = None
    count: int = 0
    error: str | None = None


class ExonResponse(BaseModel):
    """Response schema for single exon operations"""
    success: bool
    data: ExonDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/omics/exon.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const ExonSchema = z.object({
  id: z.number(),
  transcript_id: z.number().int(),
  ensembl_exon_id: z.string().min(1).max(50),
  start_position: z.number().int(),
  end_position: z.number().int(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateExonSchema = z.object({
  transcript_id: z.number().int(),
  ensembl_exon_id: z.string().min(1).max(50),
  start_position: z.number().int(),
  end_position: z.number().int(),
});

export const UpdateExonSchema = z.object({
  transcript_id: z.number().int().optional(),
  ensembl_exon_id: z.string().min(1).max(50).optional(),
  start_position: z.number().int().optional(),
  end_position: z.number().int().optional(),
});

// ==========================================
// Types
// ==========================================

export type Exon = z.infer<typeof ExonSchema>;
export type CreateExon = z.infer<typeof CreateExonSchema>;
export type UpdateExon = z.infer<typeof UpdateExonSchema>;

// ==========================================
// Response Types
// ==========================================

export type ExonResponse = {
  success: boolean;
  data?: Exon;
  error?: string;
};

export type ExonsResponse = {
  success: boolean;
  data?: Exon[];
  count?: number;
  error?: string;
};
